[Team-18][BE][타니&검봉] - 3주차 첫번째 PR

crispindeity opened the pull request:
안녕하세요! team-18 백엔드 타니&검봉입니다!

3주차 첫번째 PR요청 드립니다.

항상 저희에 부족한 코드에 좋은 리뷰를 남겨주셔서 감사합니다.

이번에도 많은 채찍과 당근 부탁드립니다^^

### 작업 내용
- 2주차 마지막 PR리뷰 반영
- 숙소 검색 API 개발 (도시, 체크인, 체크아웃, 최소금액, 최대금액, 총 게스트 수용 가능 인원수 필터링)
- [현재 개발된 API](https://github.com/bn-tw2020/airbnb/wiki/AirBnb-%EC%88%99%EC%86%8C-%EC%98%88%EC%95%BD-%EC%84%9C%EB%B9%84%EC%8A%A4-API-%EC%B4%88%EC%95%88-%EC%84%A4%EA%B3%84)

### 궁금한 점
- 검색을 하기위해 사용되는 `SearchController`에서 현재 `RegionService`와 `SearchService`를 주입 받아 사용중인데, 이렇게 하나의 컨트롤러에서 다른 영역의 서비스(RegionService)를 주입받아 사용하는게 잠재적으로 문제를 일으킬 가능성이 있을지 궁금합니다.
- 숙소의 지역을 필터링 하기 위해 `RequestParam`으로 도시의 이름을 받아서, `RegionService`에서 도시의 이름을 바탕으로 `Region`의 아이디를 찾아 `SearchController`에서 사용중인데, 현재로서는 이런 방법밖에 생각이 나질 않는데 Region 아이디을 알기위해 DB에 접근하는 방식이 맞는 방식인지 아니면 DB 설계를 잘못하여 이런 방법밖에 없는것인지 궁금합니다.
- `SearchRepository`에서 현재 `native query`를 사용하고 있는데, 타니의 생각으로는 95%이상 JPQL로 작성해도 무방한것으로 알고 있습니다.  혹시 아래의 쿼리를 JPQL로 변환이 가능할지 궁금합니다.

```sql
select *
from accommodation as a
where a.id NOT IN
      (select accommodation_id
       from schedule
       where stay_date between :checkinTime and :checkoutTime)
and a.region_id = :regionId
and a.amount_of_day between :minPrice and :maxPrice
and a.max_guest <= :maxGuest;
```

juni8453 committed:
chore: 변수명 수정 (44)
- 관용적인 표현이라도 줄임말을 지양한다.

juni8453 committed:
feat: User, Reservation 엔티티 매핑의 주인 변경 (44)
- OneToMany 에서 ManyToOne 으로 Reservation 이 주인이 될 수 있도록 변경

juni8453 committed:
feat: Reservation, Accommodation 엔티티 양방향 -> 단방향 수정 (44)

juni8453 committed:
feat: LocalDateTime -> LocalDate 로 변경 (45)

checkIn, checkOut 은 시간 개념이 필요없기 때문에 LocalDate 로 수정

juni8453 committed:
feat: data.sql 추가 (45)

테스트용 데이터 삽입

juni8453 committed:
feat: Entity 추가, 수정 및 Enum 객체 수정 (45)

1. Accommodation Entity 수정
  - AccommodationFee 임베디드 타입 추가

2. AccommodationFee 임베디드 수정
  - 예약 시 숙박 가격의 총합을 구해내는 계산 메소드 추가

3. Schedule Entity 추가
  - 숙소 내부의 예약 상태를 확인하는 Entity

4. Reservation Entity 추가
  - maxPrice, minPrice 필드 값 삭제
  - totalAmountReservation 필드 값 추가

4. DiscountPolicy 수정
  - 주 단위 금액 퍼센트를 사용하도록 getter 추가

juni8453 committed:
feat: 날짜 (체크인, 체크아웃) 에 의한 숙소 검색 필터링 구현 (45)

- checkinTime, checkoutTime 을 클라이언트를 통해 받아서, 해당 날짜 사이에 예약이 잡혀있지 않은 숙소 리스트를 반환하도록 구현

crispindeity committed:
feat: 숙소 날짜 구간으로 조회하는 기능 추가(45)

- 숙소 검색 시 해당 날짜에 예약이 잡혀 있으면 검색 되지 않도록 하는 기능 구현
- 더미데이터 추가

juni8453 committed:
feat: 최소, 최대가격과 최대 게스트 수에 의한 숙소 검색 필터링 구현 (45)

- 최소, 최대 가격 및 최대 게스트 수로 필터링 후 필터되지 않은 숙소 리스트를 반환하도록 구현

crispindeity committed:
feat: 숙소 지역 검색 기능 구현(41)

- 숙소 검색 시 지역을 구별하여 검색되도록 구현

crispindeity committed:
chore: Dto Converter 수정(50)

- Service에 있던 Dto Converter를 Controller로 이동

crispindeity committed:
chore: @Column 제거(50)

- PK값에 붙어 있던 @Column 어노테이션 제거

crispindeity committed:
chore: 코드 리포맷(50)

- 코드 리포맷팅

juni8453 committed:
배포를 위한 application-deploy.yml 파일 수정

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/Accommodation.java:
```diff
@@ -40,26 +38,23 @@ public class Accommodation {
         this.description = description;
         this.name = name;
-        this.mainImgUrl = mainImgUrl;
+        this.mainImageUrl = mainImgUrl;
```
파라미터 변수도 같이 확인해주세요!

crispindeity replied:
넵 좀 더 꼼꼼하게 살펴보는 습관을 들이겠습니다 감사합니다!

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/AccommodationFee.java:
```diff
@@ -0,0 +1,57 @@
+    private int roomCharge;
+
+    // 총 예약 확정 금액
+    private int totalAmountOfReservation;
```
1. AccomodationFee는 Accomodation테이블에 embedded 되어서 들어갈 데이터이기 때문에 숙박관련데이터만 존재해야할 것 같아요. 
totalAmoutOfDays와 totalAmountOfReservation 필드는 예약관련데이터라고 볼 수 있으니 여기에 이으면 안되는 데이터라고 봐야겠죠?

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/DiscountPolicy.java:
```diff
@@ -8,4 +8,8 @@ public enum DiscountPolicy {
 
     private final double percent;
+
+    public double getDiscountPolicy() {
+        return percent;
+    }
```
```suggestion
    public double getDiscountRate() {
        return discountRate;
    }
```

함수명과 필드명을 위처럼 바꾸는게 어떨지 제안드려봅니다!

crispindeity replied:
넵 제안 감사합니다.
제안해 주시것이 더 직관적인거 같네요 ㅎ.ㅎ

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/Region.java:
```diff
@@ -15,7 +15,6 @@ public class Region {
 
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
-    @Column(name = "region_id")
```
👍

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/Reservation.java:
```diff
@@ -16,14 +15,13 @@ public class Reservation {
 
     @Column(nullable = false)
-    private LocalDateTime checkInTime;
+    private LocalDate checkinTime;
```
LocalDate로 다시 바꾼 이유가 있을까요?

crispindeity replied:
체크인, 체크아웃에서 시간을 사용하지 않아 변경하였습니다.
생각해보니 필드명 또한 Time이 포함될 필요가 없는거 같습니다. 수정하도록 하겠습니다.

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/Reservation.java:
```diff
@@ -35,42 +33,31 @@ public class Reservation {
 
     @Column(nullable = false)
-    private int maxPrice;
+    private double totalAmountReservation;
```
totalAmount 정도로 두어도 괜찮을것 같네요 :)

crispindeity replied:
넵 알겠습니다 ㅎㅎ

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/AccommodationFee.java:
```diff
@@ -0,0 +1,57 @@
+    private int totalAmountOfReservation;
+
+    // 주 단위 요금 할인
+    @Enumerated(EnumType.STRING)
+    private DiscountPolicy discountPolicy;
```
DiscountPolicy가 어떤용도로 사용되고 있는지 직관적이지 않은 것 같아요
이것도 예약데이터인지, 숙박 데이터인지 같이 구별해주세요!

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/AccommodationFee.java:
```diff
@@ -0,0 +1,57 @@
+        long stayDays = diff.toDays();
+
+        AccommodationFee accommodationFee = accommodation.getAccommodationFee();
+        double discountPolicyOfWeek = DiscountPolicy.WEEK.getDiscountPolicy();
```
가격 계산시 DiscountPolicy를 Week를 하드코딩으로 사용하는데 의도하신부분이 맞을까요?
(맞다면 부연설명 부탁드릴게요!)

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/AccommodationFee.java:
```diff
@@ -0,0 +1,57 @@
+    @Enumerated(EnumType.STRING)
+    private DiscountPolicy discountPolicy;
+
+    public double calculationTotalAmountOfDay(LocalDate checkinTime, LocalDate checkoutTime, Accommodation accommodation) {
```
1. 메서드의 이름의 시작은 가급적 동사로 시작하도록 해주세요!
2. accomodationFee 안에 정의된 메서드이니 accomodation을 굳이 받아서 다시 accomodationFee를 꺼낼 필요가 없겠죠?

```suggestion
    public double calculateTotalAmount(LocalDate checkinTime, LocalDate checkoutTime) {
```

crispindeity replied:
넵 알겠습니다. 바로 변경하겠습니다.
다른 클래스에서 계산을 사용하는 로직을 그대로 복붙하여, Accommodation이 포함된거 같습니다.
중복된 로직이라 합치고 싶지만, 현재 DB 설계상 나눌수 밖에 없어 저희도 고민을 좀 더 해보도록 하겠습니다.

c-yeonwoo replied:
db 설계를 논외하고도 Accomodation 없이 계산이 가능할것같아요 :)

c-yeonwoo commented on BE/src/main/java/team18/airbnb/domain/Schedule.java:
```diff
@@ -0,0 +1,24 @@
+import javax.persistence.*;
+import java.time.LocalDate;
+
+@Entity
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Schedule {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
```
이건 어떤 용도(역할)로 만들어진 엔티티일까요?

c-yeonwoo commented on BE/src/main/java/team18/airbnb/region/RegionController.java:
```diff
@@ -24,7 +25,11 @@ public class RegionController {
 
         List<AccommodationByConceptDto> accommodationByConceptDtos = regionService.createAccommodationByConceptDto();
-        List<LookAroundRegionDto> lookAroundRegionDtos = regionService.createLookAroundRegionDto();
+        List<Region> regions = regionService.createLookAroundRegion();
```
1. 여긴 url 이름과 맞게 mainController 또는 homeController가 더 맞을 것 같군요.  (region이 search 영역에서도 사용되고 있는 도메인이기 때문에 분리하는게 좋을 것 같아요)
2. createAccomodationByConceptDto 데이터를 현재는 하드코딩해서 사용하고 있는데, 혹 임시용이 아니라면 이 데이터들도 db로 관리하는게 좋겠죠? 그렇게 되면 accomodationByConcept 이라는 이름보다 Suggetion, Recommendation 같은용어가 더 적절하지 않을까 제안드려 볼게요! regionService에서 가져오는것도 다시 생각해봐야겠죠?
<img width="357" alt="image" src="https://user-images.githubusercontent.com/42382027/172951864-c13dd509-4de3-4548-ac27-ee0faf1edfd0.png">
3. 2번의 내용과 함께 create이라는 메서드 명보다는 조회 역할을 하기 때문에 find~, read~ 로 메서드명을 지어주시는게 더 자연스럽지않을까요?

c-yeonwoo commented on BE/src/main/java/team18/airbnb/region/RegionService.java:
```diff
@@ -32,11 +30,13 @@ public List<AccommodationByConceptDto> createAccommodationByConceptDto() {
+    public List<Region> createLookAroundRegion() {
+        List<Region> allRegion = regionRepository.findAll();
 
-            return allRegion.stream()
-                    .map(LookAroundRegionDto::new)
-                .collect(Collectors.toList());
+        return allRegion;
+    }
```
search API에서도 이 메서드를 그대로 쓰는것 같은데, findAll을 하게 되면 region데이터가 많아질수록 시간이 오래걸리겠죠? 지금은 데이터가 없으니 괜찮지만, 추후엔 페이징을 고려하거나, limit을 걸어서 상위 몇개 데이터만 보여주는게 좋을 것 같네요 :)

c-yeonwoo commented on BE/src/main/java/team18/airbnb/search/SearchService.java:
```diff
@@ -1,11 +1,21 @@
 
     private final SearchRepository searchRepository;
+
+    public List<Accommodation> findTest(Long regionId, LocalDate checkinTime, LocalDate checkoutTime, int minPrice, int maxPrice, int maxGuest) {
```
메서드명 한번 확인해주세요~

crispindeity replied:
넵 바로 변경하도록 하겠습니다!

c-yeonwoo commented on BE/src/main/java/team18/airbnb/search/SearchService.java:
```diff
@@ -1,11 +1,21 @@
+
+    public List<Accommodation> findTest(Long regionId, LocalDate checkinTime, LocalDate checkoutTime, int minPrice, int maxPrice, int maxGuest) {
+
+        return searchRepository.findFilterResultAccommodations(regionId, checkinTime, checkoutTime, minPrice, maxPrice, maxGuest);
```
메서드명이 불필요하게 길지 않아도 될것 같아요. 
`findAccomodationsBy` 정도로만 정의해도 좋을것 같네요

crispindeity replied:
항상 좋은 제안 너무나 감사드립니다! ㅎㅎ

c-yeonwoo commented on BE/src/main/java/team18/airbnb/search/dto/RegionNameDto.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.Getter;
+import team18.airbnb.domain.Region;
+
+@Getter
+public class RegionNameDto {
+
+    private final Long id;
+    private final String regionName;
+
+    public RegionNameDto(Region region) {
+        this.id = region.getId();
+        this.regionName = region.getRegionName();
+    }
+}
```
조인쿼리가 작성된다면, 불필요한 dto가 되겠군요 !

c-yeonwoo commented on BE/src/main/java/team18/airbnb/search/SearchController.java:
```diff
@@ -1,24 +1,61 @@
+            @RequestParam("maxPrice") int maxPrice,
+            @RequestParam("maxGuest") int maxGuest) {
+
+        Region region = regionService.findRegionByRegionName(regionName);
+        RegionNameDto regionNameDto = new RegionNameDto(region);
+
+        Long regionId = regionNameDto.getId();
+
+        List<Accommodation> accommodations = searchService.findTest(regionId, checkinTime, checkoutTime, minPrice, maxPrice, maxGuest);
+
+        List<SearchResultDto> searchResultDtos = accommodations.stream()
+                .map(accommodation -> new SearchResultDto(checkinTime, checkoutTime, accommodation))
```
> * 숙소의 지역을 필터링 하기 위해 RequestParam으로 도시의 이름을 받아서, RegionService에서 도시의 이름을 바탕으로 Region의 아이디를 찾아 SearchController에서 사용중인데, 현재로서는 이런 방법밖에 생각이 나질 않는데 Region 아이디을 알기위해 DB에 접근하는 방식이 맞는 방식인지 아니면 DB 설계를 잘못하여 이런 방법밖에 없는것인지 궁금합니다.

이럴때 Region테이블과 Accomodation 테이블을 조인하면 쿼리를 두번 수행하지 않아도 되겠죠?

c-yeonwoo commented on BE/src/main/java/team18/airbnb/search/SearchRepository.java:
```diff
@@ -1,7 +1,32 @@
+
+public interface SearchRepository extends JpaRepository<Accommodation, Long> {
+
+    @Query(value = "select * " +
+            "from accommodation as a " +
+            "where a.id NOT IN " +
+            "(select accommodation_id " +
+            " from schedule where stay_date between :checkinTime and :checkoutTime)" +
+            " and a.region_id = :regionId" +
+            " and a.amount_of_day between :minPrice and :maxPrice" +
+            " and a.max_guest <= :maxGuest ", nativeQuery = true)
```
> * SearchRepository에서 현재 native query를 사용하고 있는데, 타니의 생각으로는 95%이상 JPQL로 작성해도 무방한것으로 알고 있습니다. 혹시 아래의 쿼리를 JPQL로 변환이 가능할지 궁금합니다.
select *
from accommodation as a
where a.id NOT IN
      (select accommodation_id
       from schedule
       where stay_date between :checkinTime and :checkoutTime)
and a.region_id = :regionId
and a.amount_of_day between :minPrice and :maxPrice
and a.max_guest <= :maxGuest;

물론 JPQL로 작성이 가능합니다.! 하지만 위에서 언급드렸듯이 Region테이블을 조인해야하기 때문에 쿼리가 좀 더 복잡해질 수 있는데요 :) 조인쿼리까지 한번 같이 고민해보시죠
다만 여기서는 현재 모든 파라미터들이 꼭 들어와야만 조회가 가능한 구조로 되어있는데요. 
보통 검색을 구현할 때는 요청으로 들어온 파라미터만 사용해서 선택적으로 where문을 만드는 부분이 중요합니다.
그래서 이를 쉽게 지원해주는 querydsl을 사용하는데, 이 부분에 대한 학습과 함께 구현이 되면 아주 좋을 것 같습니다. !

c-yeonwoo reviewed (approved):
> * 검색을 하기위해 사용되는 SearchController에서 현재 RegionService와 SearchService를 주입 받아 사용중인데, 이렇게 하나의 컨트롤러에서 다른 영역의 서비스(RegionService)를 주입받아 사용하는게 잠재적으로 문제를 일으킬 가능성이 있을지 궁금합니다.

컨트롤러에서 다른 영역의 서비스를 의존하는 것은 자연스러운 일입니다. 
다만 의존의 방향은 항상 Controller > Service > Repository 만 흘러야하고,
Service끼리는 의존가능하다는 점 참고해주시면 좋을 것 같아요 :)
고생하셨습니다!

crispindeity committed:
feat: 3주차 리뷰 일부 수정

- [x] 축약된 파라미터 변수명 변경
- [x] 어색한 함수명과 필드명 변경
- [x] 동사로 시작하지 않는 메서드명 수정
- [x] RegionController클래스 이름 변경

c-yeonwoo commented:
혹시 해당 브랜치에서 작업중이셨다면 revert 후 진행해주세요 :) !

juni8453 commented:
항상 리뷰 감사합니다 ! 저희가 이번 주말에는 작업을 하지 않기도 하고 다음 프로젝트 때 또한 검봉과 함께하게 돼서,  말씀해주신 것들 위주로 엔티티 매핑에 대해 더 생각 해보고 공부한 뒤 내일부터 새로 시작하는 프로젝트에서 적용할 수 있도록 해보겠습니다. 3주 동안 값진 리뷰 감사했습니다 ! ☺️

crispindeity commented:
@yeonnseok 3주 동안 해주신 리뷰를 바탕으로 많은걸 배울 수 있어 너무 감사합니다! 이번 프로젝트를 진행하면서 배운걸 바탕으로 타니와 다음 프로젝트에 적용해 보도록 하겠습니다.😄
